File: benches/comparative/typescript/bench_ws_diy.mjs
// WebSocket hot-path bench (TypeScript): OpenPX-equivalent vs DIY.
//
// The official Polymarket and Kalshi TS SDKs (@polymarket/clob-client,
// kalshi-typescript-sdk) don't ship WebSocket. A user replicating
// real-time book state writes: ws subscribe + JSON.parse + Map-based
// book. This bench measures the cost of that DIY path on real frames.
//
// The OpenPX-side number is sourced from the Rust `ws_decode_apply`
// bench (cargo bench -p px-bench-comparative --bench hot_path). The
// Node user pays roughly that plus ~50-200ns of NAPI marshalling per
// message.
//
// Writes JSON summary to stdout.
//
// Run: `node bench_ws_diy.mjs > ../results/typescript_ws_diy.json`

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { Bench } from "tinybench";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const FIXTURE = path.join(
  __dirname,
  "..",
  "fixtures",
  "polymarket_ws_book.jsonl"
);

if (!fs.existsSync(FIXTURE)) {
  console.error(`missing fixture ${FIXTURE}; run capture_bench_fixtures.py`);
  process.exit(1);
}

const lines = fs.readFileSync(FIXTURE, "utf8").split("\n");
// Skip array-wrapped snapshot frames so the bench matches the Rust
// `ws_decode_apply` set exactly.
const FRAMES = lines.filter((l) => l && l[0] !== "[");

function diyDecodeApply(frames) {
  const bids = new Map();
  const asks = new Map();
  let processed = 0;
  for (const frame of frames) {
    const msg = JSON.parse(frame);
    const etype = msg.event_type;
    if (etype === "book") {
      bids.clear();
      asks.clear();
      for (const lvl of msg.bids ?? []) bids.set(lvl.price, lvl.size);
      for (const lvl of msg.asks ?? []) asks.set(lvl.price, lvl.size);
    } else if (etype === "price_change") {
      for (const ch of msg.price_changes ?? []) {
        const side = (ch.side ?? "").toUpperCase();
        const book = side === "BUY" ? bids : asks;
        const sz = ch.size ?? "0";
        const px = ch.price ?? "";
        if (!px) continue;
        if (parseFloat(sz) === 0) book.delete(px);
        else book.set(px, sz);
      }
    }
    // last_trade_price has no book impact in the minimal handler
    processed += 1;
  }
  return { processed, levels: bids.size + asks.size };
}

const bench = new Bench({ time: 0, iterations: 20 });
bench.add("diy::polymarket::ws_decode_apply", () => {
  diyDecodeApply(FRAMES);
});

await bench.run();

const tasks = bench.tasks.map((t) => ({
  name: t.name,
  mean_ns: t.result?.mean ? t.result.mean * 1e6 : null,
  stddev_ns: t.result?.sd ? t.result.sd * 1e6 : null,
  samples: t.result?.samples?.length ?? 0,
  msgs_per_run: FRAMES.length,
}));

console.log(JSON.stringify({ tasks }, null, 2));

File: benches/comparative/python/bench_ws_diy.py
"""WebSocket hot-path bench (Python): OpenPX-equivalent vs DIY.

The official Polymarket and Kalshi Python SDKs don't ship WebSocket
support. A user who wants real-time data has to roll their own client:
connect, subscribe, parse JSON, maintain orderbook state. This bench
measures that DIY cost on real captured frames.

It's not strictly head-to-head against an "SDK" — there is no SDK.
The comparison is "OpenPX gives you this for free; here's what
you'd pay to write the same thing yourself in pure Python."

The OpenPX-side number is sourced from the Rust hot-path bench
(`cargo bench -p px-bench-comparative --bench hot_path` group
`ws_decode_apply::openpx`) — OpenPX's WS hot path runs in Rust under
the FFI, so the Python user pays roughly that cost plus a few µs of
PyO3 marshalling per message.

Run: `pytest benches/comparative/python/bench_ws_diy.py --benchmark-only`
"""

from __future__ import annotations

import json
from pathlib import Path

import pytest

FIXTURE = (
    Path(__file__).resolve().parents[1] / "fixtures" / "polymarket_ws_book.jsonl"
)


def _load_frames() -> list[bytes]:
    """Read the JSONL fixture once at module-load. Skip array-wrapped
    snapshot frames so DIY and OpenPX bench the same frame set."""
    if not FIXTURE.exists():
        pytest.skip(f"missing fixture {FIXTURE}; run capture_bench_fixtures.py")
    out: list[bytes] = []
    with FIXTURE.open("rb") as fh:
        for line in fh:
            line = line.rstrip(b"\n")
            if not line or line[:1] == b"[":
                continue
            out.append(line)
    return out


FRAMES = _load_frames()


def _diy_decode_apply(frames: list[bytes]) -> tuple[int, int]:
    """A representative hand-rolled Python WS handler — what a user would
    write to replace OpenPX. `json.loads` + dict-per-side orderbook
    state, with delta application by event_type. Roughly 30 lines, no
    framework. Returns `(messages_processed, levels_after)`."""
    bids: dict[str, str] = {}
    asks: dict[str, str] = {}
    processed = 0
    for frame in frames:
        msg = json.loads(frame)
        etype = msg.get("event_type")
        if etype == "book":
            bids = {lvl["price"]: lvl["size"] for lvl in msg.get("bids", [])}
            asks = {lvl["price"]: lvl["size"] for lvl in msg.get("asks", [])}
        elif etype == "price_change":
            for ch in msg.get("price_changes", []):
                side = (ch.get("side") or "").upper()
                book = bids if side == "BUY" else asks
                size = ch.get("size") or "0"
                price = ch.get("price") or ""
                if not price:
                    continue
                if float(size) == 0.0:
                    book.pop(price, None)
                else:
                    book[price] = size
        # last_trade_price has no book impact in this minimal handler
        processed += 1
    return processed, len(bids) + len(asks)


def test_diy_ws_decode_apply(benchmark):
    """DIY Python WS handler: json.loads + dict state machine.
    20 rounds, deterministic on captured bytes — no network involved.
    The README cites this directly."""
    if not FRAMES:
        pytest.skip("no frames")
    result = benchmark.pedantic(
        _diy_decode_apply,
        args=(FRAMES,),
        iterations=1,
        rounds=20,
    )
    assert result[0] > 0
